### PR TITLE
Allow for single grype installation with multiple shared DBs

### PR DESCRIPTION
**Pull request by wagoodman**
This PR allows for a single installation of grype share multiple DB and track by the checksum of the DB (or is as `oss`).

It was noticed while working on this that:
- the root yardstick functions (such as `compare_results`) do not allow for passing an optional `store_root` as done elsewhere. This option has been added.
- some adjustments to logging where done to shorten the output, such as bump some log lines from info to debug ~and to summarize long tool and image strings intelligently~ (edit: descoped)
- add debug logging when hitting negative cases for capturing a result
- note explicitly in logging when a scan configuration is being replaced in a result set
- in the files I modified I tried to make the type hint annotations more accurate

### CODE REVIEW

File: src/yardstick/tool/syft.py [hunk 1]
```diff
@@ -43,9 +44,13 @@ def _install_from_installer(
         if not path:
             path = tempfile.mkdtemp()
             atexit.register(shutil.rmtree, path)
-        else:
-            if os.path.exists(os.path.join(path, "syft")):
-                tool_exists = True
+
+        if add_version_to_path:
```
Comment (willmurphyscode) [inline src/yardstick/tool/syft.py:48]: I don't understand the point of this variable. Why would we not want to add the version to the path? I thought that was how yardstick kept track of whether it needed to reinstall.

File: src/yardstick/tool/grype.py [hunk 2]
```diff
@@ -160,18 +171,29 @@ def _install_from_path(
...
         if not path:
             path = tempfile.mkdtemp()
             atexit.register(shutil.rmtree, path)
+
+        # get the description and head ref from the repo
+        src_repo_path = os.path.abspath(os.path.expanduser(src_path))
+        build_version = utils.local_build_version_suffix(src_repo_path)
+
+        if add_version_to_path:
+            path = os.path.join(path, build_version)
```
Comment (willmurphyscode) [inline src/yardstick/tool/grype.py:190]: Doesn't this already happen on line 194?

File: src/yardstick/tool/grype.py [hunk 3]
```diff
@@ -83,11 +90,12 @@ def _install_from_installer(
         return Grype(path=path, version_detail=version, **kwargs)
 
     @classmethod
-    def _install_from_git(
+    def _install_from_git(  # pylint: disable=too-many-branches
```
Comment (willmurphyscode) [inline src/yardstick/tool/grype.py:93]: There are a lot of `pylint: disable` in here. Is it time to refactor?

File: src/yardstick/store/tool.py [hunk 4]
```diff
@@ -10,15 +10,26 @@
...
+
+
 def install_path(config: artifact.ScanConfiguration, store_root: str | None = None) -> str:
     if not store_root:
         store_root = store_config.get().store_root
 
+    store_base = install_base(name=config.tool_name, store_root=store_root)
+
+    version = config.tool_version
+    if config.tool_name.startswith("grype"):
```
Comment (willmurphyscode) [inline src/yardstick/store/tool.py:27]: The `tool` base class knowing about grype is another signal that it might be time to refactor.

File: src/yardstick/cli/config.py [hunk 5]
```diff
@@ -68,6 +68,9 @@ class ResultSet:
     declared: list[artifact.ScanRequest] = field(default_factory=list)
     matrix: ScanMatrix = field(default_factory=ScanMatrix)
 
+    def images(self) -> list[str]:
```
Comment (willmurphyscode) [inline src/yardstick/cli/config.py:71]: I don't understand this change. Is this used somewhere else in the PR?

File: src/yardstick/capture.py [hunk 6]
```diff
@@ -111,7 +125,8 @@ def one(
     return scan_config
 
 
-def result_set(  # pylint: disable=too-many-locals
+# pylint: disable=too-many-branches, too-many-locals
```
Comment (willmurphyscode) [inline src/yardstick/capture.py:128]: This seems like signal that it might be time to refactor.

File: src/yardstick/capture.py [hunk 7]
```diff
@@ -81,12 +81,26 @@ def intake(config: artifact.ScanConfiguration, raw_results: str) -> artifact.Sca
     return artifact.ScanResult(config=config, metadata=metadata, **keys)
 
 
+def summarize_image(image: str):
```
Comment (willmurphyscode) [inline src/yardstick/capture.py:84]: Are this method and `summarize_tool` below meant to be used somewhere? I didn't see any references, but maybe I missed something.

File: src/yardstick/tool/grype.py [hunk 8]
```diff
@@ -286,31 +331,38 @@ def install(
...
-            logging.debug("updating db")
+            logging.debug("updating db from OSS")
             tool_obj.run("db", "update")
 
         return tool_obj
 
     @property
     def db_root(self) -> str:
-        return os.path.join(self.path, "db")
+        return os.path.join(self.path, "db", self.db_identity)
```
Comment (willmurphyscode) [inline src/yardstick/tool/grype.py:365]: So this will just have `oss` in it for grype installs that are using the normal database?

File: src/yardstick/store/scan_result.py [hunk 9]
```diff
@@ -77,7 +77,7 @@ def find(
     by_tool: str = "",
     by_time: str = "",
     by_description: str = "",
-    store_root: str = None,
+    store_root: Optional[str] = None,
```
Comment (willmurphyscode) [inline src/yardstick/store/scan_result.py:80]: nit: I'd prefer if unrelated changes were in a separate PR.

**Comment by wagoodman** [comment]
Let me take another stab at this one. Thanks for the review @willmurphyscode

File: src/yardstick/capture.py (hunk 7 above)
Comment (wagoodman) [reply src/yardstick/capture.py:84]: yeah, the rebase got a little rough. I'll put it back

Comment (wagoodman) [reply src/yardstick/capture.py:84]: I'll yank it entirely from the PR

File: src/yardstick/capture.py (hunk 6 above)
Comment (wagoodman) [reply src/yardstick/capture.py:128]: since I don't have the time now to do that, I'll remove the log lines

File: src/yardstick/cli/config.py (hunk 5 above)
Comment (wagoodman) [reply src/yardstick/cli/config.py:71]: it's being used in grype-db, but I can add the helper there (and remove it here)

File: src/yardstick/store/scan_result.py (hunk 9 above)
Comment (wagoodman) [reply src/yardstick/store/scan_result.py:80]: I highlighted that I was upgrading logging and type hints around store_root as found while integrating this into grype-db

File: src/yardstick/store/tool.py (hunk 4 above)
Comment (wagoodman) [reply src/yardstick/store/tool.py:27]: one of a hundred signals 🔥

File: src/yardstick/tool/grype.py (hunk 3 above)
Comment (wagoodman) [reply src/yardstick/tool/grype.py:92]: I've removed the need for the pylint exception with recent changes

File: src/yardstick/tool/grype.py (hunk 8 above)
Comment (wagoodman) [reply src/yardstick/tool/grype.py:328]: yup 👍

File: src/yardstick/tool/syft.py (hunk 1 above)
Comment (wagoodman) [reply src/yardstick/tool/syft.py]: removed

File: src/yardstick/tool/grype.py (hunk 2 above)
Comment (wagoodman) [reply src/yardstick/tool/grype.py]: removed

File: src/yardstick/cli/config.py (hunk 5 above)
Comment (willmurphyscode) [reply src/yardstick/cli/config.py:71]: Thanks! I didn't realize it was considered part of the exported library API.

**Review by willmurphyscode — APPROVED** [review]

File: src/yardstick/store/scan_result.py (hunk 9 above)
Comment (willmurphyscode) [reply src/yardstick/store/scan_result.py:80]: I saw that. I just wish it had all been done as a previous PR; that would have made it much easier to review both PRs.
